java-connectors/kafka-connect-azure-servicebus/src/main/java/io/lenses/streamreactor/connect/azure/servicebus/source/TaskToReceiverBridge.java

```diff
@@ -0,0 +1,156 @@
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.stream.Collectors;
+import lombok.AllArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.kafka.connect.source.SourceRecord;
+import org.apache.kafka.connect.storage.OffsetStorageReader;
+
+/**
+ * Bridge between Receivers and Connector's Task for Azure Service Bus.
+ */
+@Slf4j
+public class TaskToReceiverBridge {
+
+  private static final int INITIAL_RECORDS_TO_COMMIT_SIZE = 500;
+  private final BlockingQueue<ServiceBusMessageHolder> recordsQueue;
+  private final ServiceBusPartitionOffsetProvider offsetProvider;
+  private final ExecutorService sentMessagesExecutors;
+  private Map<String, ServiceBusReceiverFacade> receivers;
+  private final Map<String, ServiceBusMessageHolder> recordsToCommitMap;
+
+  /**
+   * Creates Bridge between Receivers and Connector's Task for Azure Service Bus.
+   * 
+   * @param properties          all properties for {@link org.apache.kafka.connect.source.SourceTask}.
+   * @param recordsQueue        records queue used to store received messages.
+   * @param offsetStorageReader offset storage reader from Task.
+   */
+  public TaskToReceiverBridge(Map<String, String> properties,
+      BlockingQueue<ServiceBusMessageHolder> recordsQueue, OffsetStorageReader offsetStorageReader) {
+    this.recordsQueue = recordsQueue;
+    this.offsetProvider = new ServiceBusPartitionOffsetProvider(offsetStorageReader);
+    String connectionString = properties.get(AzureServiceBusConfigConstants.CONNECTION_STRING);
+    List<Kcql> kcqls =
+        KcqlConfigBusMapper.mapKcqlsFromConfig(properties.get(AzureServiceBusConfigConstants.KCQL_CONFIG));
+    sentMessagesExecutors = Executors.newFixedThreadPool(kcqls.size() * 2);
+    recordsToCommitMap = new ConcurrentHashMap<>(INITIAL_RECORDS_TO_COMMIT_SIZE);
+
+    initiateReceivers(recordsQueue, kcqls, connectionString);
+  }
+
+  /**
+   * Creates Bridge between Receivers and Connector's Task for Azure Service Bus.
+   *
+   * @param connectionString    Service Bus connection string
+   * @param kcqls               list of KCQLs to initiate Receivers for.
+   * @param recordsQueue        records queue used to store received messages.
+   * @param offsetStorageReader offset storage reader from Task.
+   */
+  public TaskToReceiverBridge(String connectionString, List<Kcql> kcqls,
+      BlockingQueue<ServiceBusMessageHolder> recordsQueue, OffsetStorageReader offsetStorageReader) {
+    this.recordsQueue = recordsQueue;
+    this.offsetProvider = new ServiceBusPartitionOffsetProvider(offsetStorageReader);
+    sentMessagesExecutors = Executors.newFixedThreadPool(kcqls.size() * 2);
+    recordsToCommitMap = new ConcurrentHashMap<>(INITIAL_RECORDS_TO_COMMIT_SIZE);
```

Comment:
Let's think about this class from a dependency injection point of view - instead of having 2 constructors, with a separate one that is only accessed in test code, it might be beneficial to have a single constructor used everywhere .

I would definitely advocate moving object creation responsibilities higher up in the code hierarchy, for example it could all happen in the `SourceTask`/`SourceConnector` only.

Reply:
Then the branching will still have to happen, but will be in `SourceTask` which essentially doesn't have to know about Bridge's implementation details (that it initializes Facades).

I'd advise leaving it as-is.

Reply:
There's no branching from what I can see? Please can you explain?

I've committed a [small example of how this refactoring could be achieved](#1247 ).

The creation code can be extracted out to a `ServiceBusReceiverMapper`, which is responsible for the creation logic/wiring.  Mostly this is what `createRecievers` was doing.
The benefits of this:

- `TaskToReceiverBridge` only has a single constructor.
- `TaskToReceiverBridge` no longer has to create the receivers, single responsibility principal.  This now falls to `ServiceBusReceiverMapper`.
- `TaskToReceiverBridge` no longer has to know about the connection string.
- Test and production code can now share the same constructor.  There's no longer a special 'test' constructor.
- All fields of `TaskToReceiverBridge` can now be made final.
- Object creation can be pushed up the stack as previously described into `AzureServiceBusSourceTask`

Reply:
My bad here, I was thinking you're simply intending to move `createRecievers` to the Task. I've added it with small modifications, thanks!